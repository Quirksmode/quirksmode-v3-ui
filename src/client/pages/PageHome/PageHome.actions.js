export const FETCH_HOME_REQUEST = 'fetch_home_request';
export const FETCH_HOME_SUCCESS = 'fetch_home_success';
export const FETCH_HOME_ERROR = 'fetch_home_error';
export const fetchHomeData = () => async (dispatch, getState, api) => {
  console.log('dispatched action');
  dispatch({
    type: FETCH_HOME_REQUEST
  });

  try {
    const res = await api.get('quirksmode/v1/pages/home');
    dispatch({
      type: FETCH_HOME_SUCCESS,
      payload: res
    });
  } catch (err) {
    dispatch({
      type: FETCH_HOME_ERROR
    });
  }
};
