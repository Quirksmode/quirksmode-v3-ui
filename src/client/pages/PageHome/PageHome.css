.PageHome {
  &__colWrap {
    @media (--tablet-portrait) {
      display: flex;
      justify-content: space-between;
    }
  }

  &__col {
    @media (--tablet-portrait) {
      margin-right: 5.6%;
      width: 50%;
    }

    &:last-child {
      margin-right: 0;
    }
  }
}
