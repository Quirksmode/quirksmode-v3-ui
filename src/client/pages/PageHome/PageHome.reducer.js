import {
  FETCH_HOME_REQUEST,
  FETCH_HOME_SUCCESS,
  FETCH_HOME_ERROR
} from './PageHome.actions';

export const initialState = {
  content: {
    title: '',
    heroSlides: [],
    featuredWork: [],
    intro: null,
    latestBlogPosts: [],
    skills: [],
  },
  metadata: {},
  loading: false,
  error: false
};

export default (state = initialState, action) => {
  switch (action.type) {
    case FETCH_HOME_REQUEST:
      return {
        ...state,
        loading: true,
        error: false
      };
    case FETCH_HOME_SUCCESS:
      return {
        ...state,
        content: action.payload.content,
        metadata: action.payload.metadata,
        loading: false,
        error: false
      };
    case FETCH_HOME_ERROR:
      return {
        ...state,
        loading: false,
        error: true
      };
    default:
      return state;
  }
};
