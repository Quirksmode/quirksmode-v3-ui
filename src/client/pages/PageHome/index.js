import React from 'react';
import loadable from '@loadable/component';
import Loading from 'components/Loading/Loading';
import ErrorBoundary from 'components/ErrorBoundary/ErrorBoundary';

const Home = loadable(() => import('./PageHome'), {
  fallback: <Loading />
});

const WrappedHome = props => (
  <ErrorBoundary>
    <Home { ...props } />
  </ErrorBoundary>
);

export default WrappedHome;
