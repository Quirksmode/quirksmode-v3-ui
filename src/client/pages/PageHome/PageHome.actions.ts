import axios from 'axios';
import { ActionTypes, PageHomeData } from './PageHome.types';
import { Dispatch } from 'redux';

export interface FetchHomeDataAction {
  type:
    | ActionTypes.FETCH_HOME_REQUEST
    | ActionTypes.FETCH_HOME_SUCCESS
    | ActionTypes.FETCH_HOME_ERROR;
  payload?: PageHomeData;
}

export const fetchHomeData = () => async (dispatch: Dispatch) => {
  dispatch<FetchHomeDataAction>({
    type: ActionTypes.FETCH_HOME_REQUEST,
  });

  try {
    const res = await axios.get<PageHomeData>(
      `${process.env.CMS_URL}/wp-json/quirksmode/v1/pages/home`
    );
    dispatch<FetchHomeDataAction>({
      type: ActionTypes.FETCH_HOME_SUCCESS,
      payload: res.data,
    });
    console.log('fetchHomeData', res.data);
  } catch (err) {
    dispatch<FetchHomeDataAction>({
      type: ActionTypes.FETCH_HOME_ERROR,
    });
  }
};
