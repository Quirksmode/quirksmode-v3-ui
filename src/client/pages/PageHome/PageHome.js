import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Meta from 'components/Meta/Meta';
import PageWrapper from 'components/PageWrapper/PageWrapper';
import { fetchHomeData } from './PageHome.actions';
import PageHomeFeatured from './PageHomeFeatured/PageHomeFeatured';
import PageHomeIntro from './PageHomeIntro/PageHomeIntro';
import PageHomeSkills from './PageHomeSkills/PageHomeSkills';
import PageHomeBlog from './PageHomeBlog/PageHomeBlog';

/**
 * Home Page
 *
 * @name PageHome
 * @param {function} props.fetchHomeDataAction - Redux action to Fetch the Home Page Data
 * @param {object} props.content - The content for this Page
 * @param {object} props.metadata - The metadata for this Page
 * @param {boolean} props.loading - Flag for while the data is being fetched
 * @param {boolean} props.error - Flag for if there is an error fetching the data
 * @param {object} props.siteSettings - The Global Site Settings
 * @return {JSXElement}
 */
const PageHome = ({
  fetchHomeDataAction,
  content,
  metadata,
  loading,
  error,
  siteSettings
}) => {
  /**
   * @type {Object}
   * @property {array} content.featuredWork - For the featured work Slider
   * @property {object} content.intro - Intro content
   * @property {array} content.latestBlogPosts - Latest Blog Posts
   * @property {array} content.skills - List of skills
   */
  const {
    featuredWork,
    intro,
    latestBlogPosts,
    skills
  } = content;

  /**
   * @type {Object}
   * @property {object} siteSettings.cv - CV data from the Global Site Settings
   */
  const {
    cv
  } = siteSettings;

  /**
   * Fetch the Home Page Data via Redux, but only if the data does not already exist
   */
  useEffect(() => {
    if (!intro) fetchHomeDataAction();
  }, [fetchHomeDataAction, intro]);

  return (
    <PageWrapper error={ error } loading={ loading }>
      <div className="page PageHome">
        <Meta { ...metadata } />
        <h1 className="visuallyHidden">Welcome to Quirksmode</h1>
        { featuredWork.length > 0 && <PageHomeFeatured featuredWork={ featuredWork } /> }
        <section className="Page__section">
          <div className="Page__sectionInner grid">
            <div className="PageHome__colWrap">
              <div className="PageHome__col">
                { intro && <PageHomeIntro intro={ intro } cv={ cv } /> }
                { skills.length > 0 && <PageHomeSkills skills={ skills } /> }
              </div>
              <div className="PageHome__col">
                <PageHomeBlog
                  latestBlogPosts={ latestBlogPosts }
                  loading={ loading }
                />
              </div>
            </div>
          </div>
        </section>
      </div>
    </PageWrapper>
  );
};

PageHome.propTypes = {
  fetchHomeDataAction: PropTypes.func,
  content: PropTypes.object,
  metadata: PropTypes.object,
  loading: PropTypes.bool,
  error: PropTypes.bool,
  siteSettings: PropTypes.object
};

const mapStateToProps = ({
  pageHome,
  app
}) => ({
  content: pageHome.content,
  metadata: pageHome.metadata,
  loading: pageHome.loading,
  error: pageHome.error,
  siteSettings: app.siteSettings
});

const mapDispatchToProps = dispatch => ({
  fetchHomeDataAction: (...args) => dispatch(fetchHomeData(...args))
});

export default connect(mapStateToProps, mapDispatchToProps)(PageHome);
