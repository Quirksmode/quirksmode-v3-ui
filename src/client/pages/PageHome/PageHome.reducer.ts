import { PayloadAction, createSlice } from '@reduxjs/toolkit';
import { PageHomeState, PageHomeData } from './PageHome.types';

export const initialState: PageHomeState = {
  content: {
    title: '',
    heroSlides: [],
    featuredWork: [],
    intro: null,
    latestBlogPosts: [],
    skills: [],
  },
  metadata: null,
  loading: false,
  error: false,
};

const pageHome = createSlice({
  name: 'pageHome',
  initialState,
  reducers: {
    fetchHomeRequest: (state) => {
      state.loading = true;
      state.error = false;
    },
    fetchHomeSuccess: (state, { payload }: PayloadAction<PageHomeData>) => {
      state.content = payload.content;
      state.metadata = payload.metadata;
      state.loading = false;
      state.error = false;
    },
    fetchHomeError: (state) => {
      state.loading = false;
      state.error = true;
    },
  },
});

export default pageHome.reducer;
export const {
  fetchHomeRequest,
  fetchHomeSuccess,
  fetchHomeError,
} = pageHome.actions;
