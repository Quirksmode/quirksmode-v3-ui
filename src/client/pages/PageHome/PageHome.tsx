import React, { useEffect, useMemo } from 'react';
import { useDispatch } from 'react-redux';
import { useTypedSelector } from 'client/redux/types';
import Meta from 'components/Meta/Meta';
import PageWrapper from 'components/PageWrapper/PageWrapper';
import { fetchHomeData } from './PageHome.actions';
import PageHomeFeatured from './PageHomeFeatured/PageHomeFeatured';
import PageHomeIntro from './PageHomeIntro/PageHomeIntro';
import PageHomeSkills from './PageHomeSkills/PageHomeSkills';
import PageHomeBlog from './PageHomeBlog/PageHomeBlog';

/**
 * Home Page
 */
const PageHome: React.FC = () => {
  const dispatch = useDispatch();
  const pageHome = useTypedSelector((state) => state.pageHome);
  const app = useTypedSelector((state) => state.app);

  const { content, metadata, loading, error } = pageHome;
  const { featuredWork, intro, latestBlogPosts, skills } = content;
  const { cv } = app.siteSettings;

  /**
   * Fetch the Home Page Data via Redux, but only if the data does not already exist
   */
  useEffect(() => {
    if (!intro) {
      dispatch(fetchHomeData());
    }
  }, [intro]);

  return useMemo(
    () => (
      <PageWrapper error={error} loading={loading}>
        <div className="page PageHome">
          {metadata && <Meta {...metadata} />}
          <h1 className="visuallyHidden">Welcome to Quirksmode</h1>
          {featuredWork?.length > 0 && (
            <PageHomeFeatured featuredWork={featuredWork} />
          )}
          <section className="Page__section">
            <div className="Page__sectionInner grid">
              <div className="PageHome__colWrap">
                <div className="PageHome__col">
                  {intro && <PageHomeIntro intro={intro} cv={cv} />}
                  {skills?.length > 0 && <PageHomeSkills skills={skills} />}
                </div>
                <div className="PageHome__col">
                  {latestBlogPosts?.length > 0 && (
                    <PageHomeBlog latestBlogPosts={latestBlogPosts} />
                  )}
                </div>
              </div>
            </div>
          </section>
        </div>
      </PageWrapper>
    ),
    [app, content, error, loading, metadata]
  );
};

export default PageHome;
