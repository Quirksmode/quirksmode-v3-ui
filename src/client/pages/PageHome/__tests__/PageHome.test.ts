import { mockStore } from 'client/tests/mocks/store';
import { mountRender, reduxWrap } from 'client/tests/testUtils';
import moxios from 'moxios';
import PageHome from '../PageHome';
import PageHomeStub from './PageHome.stub.json';
import { fetchHomeData } from '../PageHome.actions';
import 'tests/mocks/intersectionObserver';

describe('PageHome', () => {
  let wrapper;
  const props = {};

  beforeEach(() => {
    moxios.install();
  });

  afterEach(() => {
    moxios.uninstall();
  });

  it('matches snapshot', () => {
    wrapper = mountRender(reduxWrap(PageHome, props));
    expect(wrapper.html).toMatchSnapshot();
  });

  it('runs the correct actions when the Home data is successfully fetched', async (done) => {
    const { dispatch, getActions } = mockStore();
    wrapper = mountRender(reduxWrap(PageHome, props));

    dispatch(fetchHomeData());

    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request
        .respondWith({
          status: 200,
          response: PageHomeStub,
        })
        .then(() => {
          const expected = [
            {
              type: 'pageHome/fetchHomeRequest',
            },
            {
              type: 'pageHome/fetchHomeSuccess',
              payload: PageHomeStub,
            },
          ];
          const actual = getActions();
          expect(actual).toEqual(expected);
          done();
        });
    });
  });

  it('runs the correct actions when the Home data is not successfully fetched', async (done) => {
    const { dispatch, getActions } = mockStore();
    wrapper = mountRender(reduxWrap(PageHome, props));

    dispatch(fetchHomeData());

    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request
        .respondWith({
          status: 500,
        })
        .then(() => {
          const expected = [
            {
              type: 'pageHome/fetchHomeRequest',
            },
            {
              type: 'pageHome/fetchHomeError',
            },
          ];
          const actual = getActions();
          expect(actual).toEqual(expected);
          done();
        });
    });
  });
});
