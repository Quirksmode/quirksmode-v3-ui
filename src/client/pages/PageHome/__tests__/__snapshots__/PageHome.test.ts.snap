// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageHome matches snapshot 1`] = `[Function]`;
